src/eval.c

```diff
@@ -3605,6 +3617,304 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    // or a single quote in the double quoted string?
+    int did_encount_escaped_double_in_double =
+	(!is_literal_string && **expr == '\\' && *(*expr + 1) == '"');
+
+    int result =
+	did_encount_double_in_single ||
+	did_encount_single_single_in_single ||
+	did_encount_single_in_double ||
+	did_encount_escaped_double_in_double;
+
+    if (did_encount_single_single_in_single || did_encount_escaped_double_in_double)
+    {
+	MB_PTR_ADV(*expr);
+    }
+
+    return result;
+}
+
+    static char_u*
+remove_all_quote_escaping(char_u *expr, int is_literal_string)
+{
+    garray_T	result;
+    char_u	*p;
+
+    ga_init2(&result, 1, 80);
+
+    for (p = expr; *p != NUL; MB_PTR_ADV(p))
+    {
+	int is_escaped_quote =
+	    (is_literal_string && *p == '\'' && *(p + 1) == '\'') ||
+	    (!is_literal_string && *p == '\\' && *(p + 1) == '"');
+	if (is_escaped_quote)
+	{
+	    ga_append(&result, *(p + 1));
+	    MB_PTR_ADV(p);
+	    continue;
+	}
+	ga_append(&result, *p);
+    }
+
+    return (char_u *) result.ga_data;
+}
+
+/*
+ * Parses a string like "${var}", and consume it onto "}".
+ *
+ * Return "{x->(type(x) is v:t_string?x:string(x)}(var)", or NULL if expr is invalid.
+ */
+    static char_u*
+enclose_template_expr_by_stringify(char_u **expr, int is_literal_string)
+{
+    char_u	*stringify_head = (char_u *) "{x -> type(x) is v:t_string ? x : string(x)}(";
+    char_u	*stringify_last = (char_u *) ")";
+    char_u	*stringify_argument;
+    char_u	*expr_orig = *expr;
+    char_u	*expr_head;
+    garray_T	result;
+    size_t	nested_blocks = 0;
+    int		is_in_quote = FALSE;
+
+    ga_init2(&result, 1, 80);
+    ga_concat(&result, stringify_head);
+
+    *expr += 2;  // ${
+    expr_head = *expr;
+
+    // While the closing '}' encounted. The '}' is neither '}' of a dict nor '}' in a string.
+    for (; !(!is_in_quote && **expr == '}' && nested_blocks == 0); MB_PTR_ADV(*expr))
+    {
+	switch (**expr)
+	{
+	    case '{':
+		if (!is_in_quote)
+		{
+		    ++nested_blocks;
+		}
+		break;
+	    case '}':
+		if (!is_in_quote)
+		{
+		    --nested_blocks;
+		}
+		break;
+	    case NUL:
+		ga_clear(&result);
+		semsg(_("E1000: Unterminated template literal: %s"), expr_orig);
+		return NULL;
+	    default: {
+		if (did_encount_string_quote(expr, is_literal_string))
+		{
+		    is_in_quote = !is_in_quote;
+		}
+	    }
+	}
+    }
+
+    if (expr_head == *expr)
+    {
+	ga_clear(&result);
+	semsg(_("E1000: Empty template literal: %s"), expr_head);
+	return NULL;
+    }
+
+    **expr = NUL;
+    stringify_argument = remove_all_quote_escaping(expr_head, is_literal_string);
+
+    ga_concat(&result, stringify_argument);
+    ga_concat(&result, stringify_last);
+
+    vim_free(stringify_argument);
+    **expr = '}';
+
+    return (char_u *) result.ga_data;
+}
+
+/*
+ * Parses a string like "$var", and consume expr to the last char of a variable name.
+ *
+ * Return "{x->(type(x) is v:t_string?x:string(x)}(var)", or NULL if expr is invalid.
```

Comment:
s/(type(x)/type(x)/

I think it is better to insert spaces for readability.
`{x -> type(x) is v:t_string ? x : string(x)}(var)`

Reply:
Thanks, I replaced the comment.